### PR TITLE
Enforce visibility struct tags

### PR DESCRIPTION
**Pull request by mbarnes**
### What this PR does

This enforces property visibility when creating or updating ARO-HCP resource types to comply with the OpenAPI spec.  The implementation encodes property visibility as struct tags.

The autorest-generated API models have the useful characteristic of all struct fields being pointers.  This allows us to identify which properties were explicitly provided in a create or update request.

However the generated models do not have property metadata encoded into struct tags, and we can't (or shouldn't) modify the generated models to add such information.

To get around this limitation, struct tags for property visibiility are encoded into the _internal_ API models, which are not generated. The internal API models are of different Go types than the generated API models and are therefore not compatible, however the field names in each set of models are identical where property visibility is explicit (as opposed to implicitly inherited from parent types).

We can therefore build a lookup table of visibility struct tags by field name to use when evaluating create and update requests using the generated API models.  We call this table a StructTagMap:

```
type StructTagMap map[string]reflect.StructTag
```

where the map keys are dotted struct field names (e.g. "A.B.C.D").

Then, while recursing over a generated API model type, we keep track of struct field names in order to build lookup keys for the table.  This effectively allows struct tags from one Go type to be grafted on to another similar but incompatible Go type.

Jira: [ARO-6347 : Validate metadata from api structs](https://issues.redhat.com/browse/ARO-6347)
Link to demo recording: 

### Special notes for your reviewer

Supporting documentation:
 - [TypeSpec Azure - Property Visibility and Other Constraints](https://azure.github.io/typespec-azure/docs/howtos/ARM/resource-type#property-visibility-and-other-constraints)
 - [Microsoft Azure REST API Guidelines - Resource Schema & Field Mutability](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#resource-schema--field-mutability)

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Deployment: The deployment process was considered and addressed if required
- [ ] Testing: New code requires new unit tests.
- [ ] Documentation: Is the documentation updated? Either in the doc located in focus area, in the README or in the code itself.
- [ ] Customers: Is this change affecting customers? Is the release plan considered?

### CODE REVIEW

**Comment by mbarnes** [comment]
@bennerv asked me (on Slack) about how we'd handle a potential field visibility change in future APIs out of concern that the metadata is on the internal structs.  I wanted to address that here.

Suppose API versions v1-v4 (I'll simplify the version IDs for the sake of discussion) all had some field "foo" that was tagged `@visibility("read", "create")` in the OpenAPI spec.

Then v5 comes along and changes field "foo" to be `@visibility("read", "create", "update")`.

Here's how I would handle that:

- Update the struct tag for field "foo" in the internal API to be `visibility:"read create update"`.  Visibility is now correct for v5 and all subsequent versions.

- Each of the API modules for v1-v4 will have their own `StructTagMap` instances.
 ```
 package v1 (or v2 or v3 or v4)

 var clusterStructTagMap = api.NewStructTagMap[api.HCPOpenShiftCluster]()
```
- We would need to override the `clusterStructTagMap` for field "foo" in each of these older API modules to restore the original visibility for field "foo".
```
func init() {
        ...

        // This field was changed in v5 to be "read create update".
        clusterStructTagMap["properties.spec.Foo"] = reflect.StructTag("visibility:\"read create\"")
}
```

**Review by mjlshen — DISMISSED** [review]
The proposed mitigation for api visibility changes makes sense to me as well.

**Review by s-amann — DISMISSED** [review]

File: internal/api/visibility.go [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+	return m
+}
+
+type validateVisibility struct {
+	m        StructTagMap
+	updating bool
+	errs     []arm.CloudErrorBody
+}
+
+func ValidateVisibility(v, w interface{}, m StructTagMap, updating bool) []arm.CloudErrorBody {
```
[the remark this answers is not in the document]
Comment (mbarnes) [reply internal/api/visibility.go:138]: But I thought Go developers love single-letter variables?  \</snark\>
Kidding.  Will make it more readable.

[the remark this answers is not in the document]
Comment (mbarnes) [reply internal/api/visibility.go]: Improved readability in a new commit.

**Review by s-amann — APPROVED** [review]

**Pull request merged**